hail/python/hail/matrixtable.py

```diff
@@ -2807,25 +2807,205 @@ def write(
+
+            display_column_widths = [
+                max(
+                    len(display_column_names[index]),
+                    len(truncated_types[index]) if should_show_types else 0,
+                    max_value_width(index),
+                )
+                for index in range(len(display_column_names))
+            ]
+
+            display_block_slices = []
+            start_index = 0
+            end_index = 1
+            block_width = display_column_widths[0] + 4 if display_column_widths else 0
+            while end_index < len(display_column_names):
+                block_width += display_column_widths[end_index] + 3
+                if block_width > width:
+                    display_block_slices.append(slice(start_index, end_index))
+                    start_index = end_index
+                    block_width = display_column_widths[end_index] + 4
+                end_index += 1
+            display_block_slices.append(slice(start_index, end_index))
+
+            def format_hline(widths):
+                if not widths:
+                    return "++\n"
+                return '+-' + '-+-'.join(['-' * width for width in widths]) + '-+\n'
+
+            def pad(value: str, width: int, right_align: bool):
+                extra_count = width - len(value)
+                if right_align:
+                    return ' ' * extra_count + value
+                else:
+                    return value + ' ' * extra_count
+
+            def format_line(values: List[str], widths: List[int], right_align: List[bool]):
+                if not values:
+                    return "||\n"
+                values = map(pad, values, widths, right_align)
+                return '| ' + ' | '.join(values) + ' |\n'
+
+            ascii_str = ''
+            is_first_display_block = True
+            for block_slice in display_block_slices:
+                if is_first_display_block:
+                    is_first_display_block = False
+                else:
+                    ascii_str += '\n'
+
+                block_display_column_widths = display_column_widths[block_slice]
+                block_right_align = right_align[block_slice]
+                hline = format_hline(block_display_column_widths)
+
+                ascii_str += hline
+                ascii_str += format_line(
+                    display_column_names[block_slice], block_display_column_widths, block_right_align
+                )
+                ascii_str += hline
+                if should_show_types:
+                    ascii_str += format_line(
+                        truncated_types[block_slice], block_display_column_widths, block_right_align
+                    )
+                    ascii_str += hline
+                for row in truncated_values:
+                    ascii_str += format_line(row[block_slice], block_display_column_widths, block_right_align)
+                ascii_str += hline
+
+            if n_rows < matrix_table.count_rows():
```

Comment:
Counting the number of rows, in general, requires visiting each row (consider `mt.explode(mt.some_array`). Instead of counting the rows, use the has_more returned value from `_take_n`.

Reply:
`has_more` is always `False` in my current implementation because of `table = matrix_table.head(n_rows, n_cols).localize_entries('entries', 'cols')`. I propose just printing the number of rows that are displayed regardless of the total row count.

Reply:
Update: I have changed my implementation to use the `has_more` value now.